File: app/src/main/java/com/mdgd/pokemon/ui/splash/state/SplashScreenEffect.java
package com.mdgd.pokemon.ui.splash.state;

import com.mdgd.mvi.states.AbstractEffect;
import com.mdgd.pokemon.ui.splash.SplashContract;

public class SplashScreenEffect extends AbstractEffect<SplashContract.View> {

    public static class LaunchWorkerEffect extends SplashScreenEffect {

        public LaunchWorkerEffect() {
        }

        @Override
        protected void handle(SplashContract.View screen) {
            screen.launchWorker();
        }
    }

    public static class ShowErrorEffect extends SplashScreenEffect {

        private final Throwable error;

        public ShowErrorEffect(Throwable e) {
            error = e;
        }

        @Override
        protected void handle(SplashContract.View screen) {
            screen.showError(error);
        }
    }

    public static class ProceedToNextScreeEffect extends SplashScreenEffect {

        public ProceedToNextScreeEffect() {
        }

        @Override
        protected void handle(SplashContract.View screen) {
            screen.proceedToNextScreen();
        }
    }
}

File: app/src/main/java/com/mdgd/pokemon/ui/splash/state/SplashScreenEffect.kt
package com.mdgd.pokemon.ui.splash.state

import com.mdgd.mvi.states.AbstractEffect
import com.mdgd.pokemon.ui.splash.SplashContract

sealed class SplashScreenEffect : AbstractEffect<SplashContract.View>() {

    class ShowError(val e: Throwable) : SplashScreenEffect() {

        override fun handle(screen: SplashContract.View) {
            screen.showError(e)
        }
    }


    object LaunchWorker : SplashScreenEffect() {

        override fun handle(screen: SplashContract.View) {
            screen.launchWorker()
        }
    }

    object NextScreen : SplashScreenEffect() {

        override fun handle(screen: SplashContract.View) {
            screen.proceedToNextScreen()
        }
    }
}
